// @zappar/zappar-threejs/lib/mesh/featurepointsmesh.js
import { FeaturePointsGeometry } from '../geometry/featurepointsgeometry.js';
import * as THREE from 'three';
const vertexShader = `

uniform mat4 pointsMatrix;
uniform float pointSize;

void main() {
  gl_PointSize = pointSize;
  gl_Position = pointsMatrix * vec4( position, 1.0 );
}
`;
const fragmentShader = `
uniform vec3 color;
uniform float opacity;

void main() {
	gl_FragColor = vec4( color, opacity );
}
`;
export class FeaturePointsMesh extends THREE.Points {
    /**
     * The size of the feature points.
     *
     * Default is `5`.
     */
    set size(size) {
        this._uniforms.pointSize.value = size;
    }
    get size() {
        return this._uniforms.pointSize.value;
    }
    /**
     * The color of the feature points.
     *
     * Default is `white`.
     */
    set color(color) {
        this._uniforms.color.value = color;
    }
    get color() {
        return this._uniforms.color.value;
    }
    /**
     * The opacity of the feature points.
     *
     * Default is `0.4`.
     */
    set opacity(opacity) {
        this._uniforms.opacity.value = opacity;
    }
    get opacity() {
        return this._uniforms.opacity.value;
    }
    constructor(worldTracker, materialProps = {}) {
        super(new FeaturePointsGeometry(), new THREE.ShaderMaterial(Object.assign({ depthTest: false, depthWrite: false, vertexShader,
            fragmentShader, blending: THREE.AdditiveBlending }, materialProps)));
        this.worldTracker = worldTracker;
        this._renderSize = new THREE.Vector2();
        this._uniforms = {
            color: { value: new THREE.Color('white') },
            opacity: { value: 0.4 },
            pointsMatrix: { value: new THREE.Matrix4() },
            pointSize: { value: 5 },
        };
        this.update = (renderer) => {
            const size = renderer.getSize(this._renderSize);
            this._uniforms.pointsMatrix.value.fromArray(this.worldTracker.computePointsDataMatrix(size.x, size.y, false));
            this.geometry.update(this.worldTracker.tracksData);
        };
        this.castShadow = false;
        this.receiveShadow = false;
        this.frustumCulled = false;
        this.material.uniforms = this._uniforms;
        this.worldTracker.tracksDataEnabled = true;
    }
    dispose() {
        this.geometry.dispose();
        this.material.dispose();
    }
}
