// @zappar/zappar-threejs/lib/mesh/planesmeshes.js
/* eslint-disable no-continue */
/* eslint-disable no-restricted-syntax */
/* eslint-disable no-underscore-dangle */
import * as Zappar from '@zappar/zappar';
import * as THREE from 'three';
import { CameraMirrorMode } from '../camera.js';
export class PlanesMeshes extends THREE.Group {
    constructor(camera, worldTracker, material) {
        super();
        this.camera = camera;
        this.worldTracker = worldTracker;
        this.material = material;
        /**
         * Emits when a plane is first intersected.
         */
        this.onIntersect = new Zappar.Event1();
        /**
         * Emits when a plane is no longer intersected.
         */
        this.onIntersectEnd = new Zappar.Event1();
        this._previousIntersections = new Set();
        this._currentIntersections = new Map();
        this._geometries = new Map();
        this._update = () => {
            var _a;
            const planes = this.worldTracker ? [...this.worldTracker.planes.values()] : [];
            const planesPresent = new Set();
            for (const plane of planes) {
                if (plane.status !== Zappar.AnchorStatus.ANCHOR_STATUS_TRACKING)
                    continue;
                planesPresent.add(plane.id);
                let existing = this._geometries.get(plane.id);
                if (!existing) {
                    const geometry = new THREE.BufferGeometry();
                    const material = this.material instanceof THREE.Material ? this.material : new THREE.MeshBasicMaterial(Object.assign({ side: THREE.DoubleSide, vertexColors: true, transparent: true, opacity: 0.2 }, ((_a = this.material) !== null && _a !== void 0 ? _a : {})));
                    existing = {
                        polygonVersion: -1,
                        geometry,
                        mesh: new THREE.Mesh(geometry, material),
                        anchorId: plane.id,
                    };
                    this.add(existing.mesh);
                    this._geometries.set(plane.id, existing);
                }
                existing.mesh.matrix.fromArray(plane.pose(this.camera.rawPose, this.camera.currentMirrorMode === CameraMirrorMode.Poses));
                existing.mesh.matrix.decompose(existing.mesh.position, existing.mesh.quaternion, existing.mesh.scale);
                if (existing.polygonVersion === plane.polygonVersion)
                    continue;
                existing.polygonVersion = plane.polygonVersion;
                const mesh = Zappar.generateMeshForPlanePolygon({
                    fade: false,
                }, plane.polygon);
                const buffer = new THREE.InterleavedBuffer(mesh.vertexBuffer, mesh.stride);
                existing.geometry.setAttribute('position', new THREE.InterleavedBufferAttribute(buffer, 3, 0));
                existing.geometry.setAttribute('normal', new THREE.InterleavedBufferAttribute(buffer, 3, mesh.normalOffset));
                existing.geometry.setAttribute('uv', new THREE.InterleavedBufferAttribute(buffer, 2, mesh.uvOffset));
                existing.geometry.setAttribute('color', new THREE.InterleavedBufferAttribute(buffer, 4, mesh.colorOffset));
                existing.geometry.setIndex(new THREE.BufferAttribute(mesh.indexBuffer, 1));
                existing.geometry.computeBoundingSphere();
                existing.geometry.computeBoundingBox();
            }
            for (const [id, plane] of this._geometries.entries()) {
                if (planesPresent.has(id))
                    continue;
                plane.mesh.removeFromParent();
                plane.geometry.dispose();
                this._geometries.delete(id);
            }
        };
        worldTracker._pipeline.onFrameUpdate.bind(this._update);
    }
    intersect(ray) {
        const ret = [];
        this._currentIntersections.clear();
        for (const anchor of this._geometries.values()) {
            anchor.mesh.updateWorldMatrix(true, false);
            const intersections = ray.intersectObject(anchor.mesh);
            if (intersections.length > 0) {
                ret.push({
                    anchorId: anchor.anchorId,
                    intersection: intersections[0],
                });
                this._currentIntersections.set(anchor.anchorId, intersections[0]);
            }
        }
        for (const [anchorId, intersection] of this._currentIntersections.entries()) {
            if (!this._previousIntersections.has(anchorId)) {
                this.onIntersect.emit({ anchorId, intersection });
            }
        }
        for (const anchorId of this._previousIntersections) {
            if (!this._currentIntersections.has(anchorId))
                this.onIntersectEnd.emit({ anchorId });
        }
        this._previousIntersections.clear();
        for (const id of this._currentIntersections.keys())
            this._previousIntersections.add(id);
        ret.sort((a, b) => a.intersection.distance - b.intersection.distance);
        return ret;
    }
}
