// @zappar/zappar-threejs/lib/index.js
import { VERSION } from './version.js';
console.log(`Zappar for ThreeJS v${VERSION}`);
export { Camera, CameraPoseMode, CameraMirrorMode, CameraProfile } from './camera.js';
export { ImageAnchorGroup } from './trackers/imageanchorgroup.js';
export { ImageTrackerLoader } from './loaders/imagetrackerloader.js';
export { InstantWorldAnchorGroup } from './trackers/instantworldanchorgroup.js';
export { FaceAnchorGroup } from './trackers/faceanchorgroup.js';
export { FaceLandmarkGroup } from './trackers/facelandmarkgroup.js';
export { FaceTrackerLoader } from './loaders/facetrackerloader.js';
export { FaceBufferGeometry } from './geometry/facebuffergeometry.js';
export { FaceMeshLoader } from './loaders/facemeshloader.js';
export { HeadMaskMesh } from './mesh/headmaskmesh.js';
export { HeadMaskMeshLoader } from './loaders/headmaskmeshloader.js';
export { LibraryLoader } from './loaders/libraryloader.js';
export { LoadingManager, DefaultLoaderUI } from './loaders/loadingmanager.js';
export { CameraEnvironmentMap } from './cameraenvironmentmap.js';
export { TargetImagePreviewBufferGeometry } from './geometry/targetimagepreviewbuffergeometry.js';
export { TargetImagePreviewMesh } from './mesh/targetimagepreviewmesh.js';
export { GroundAnchorGroup } from './trackers/groundanchorgroup.js';
export { WorldPlacementGroup } from './trackers/worldplacementgroup.js';
export { UserPlacementAnchorGroup } from './trackers/userplacementanchorgroup.js';
export { FeaturePointsGeometry } from './geometry/featurepointsgeometry.js';
export { FeaturePointsMesh } from './mesh/featurepointsmesh.js';
export { WorldAnchorGroup } from './trackers/worldanchorgroup.js';
export { ImageTracker, InstantWorldTracker, BarcodeFinder, FaceTracker, CameraSource, HTMLElementSource, onFrameUpdate, glContextSet, glContextLost, getDefaultPipeline, WorldTracker } from './defaultpipeline.js';
export { FaceMesh, FaceLandmark, FaceLandmarkName, Pipeline, LogLevel, setLogLevel, permissionDenied, permissionGranted, permissionRequest, permissionDeniedUI, permissionRequestUI, browserIncompatible, browserIncompatibleUI, loaded, loadedPromise, SequenceSource, cookiesPermitted, WorldTrackerQuality, WorldTrackerUI, WorldScaleMode, AnchorStatus, setOptions, setPreferWebXRCamera, } from '@zappar/zappar';
