// @zappar/zappar-threejs/lib/trackers/userplacementanchorgroup.js
/* eslint-disable no-underscore-dangle */
/* eslint-disable no-return-assign */
/* eslint-disable no-restricted-syntax */
/* eslint-disable max-len */
/* eslint-disable no-param-reassign */
import * as Zappar from '@zappar/zappar';
import { CameraMirrorMode } from '../camera.js';
import * as THREE from 'three';
import { PlanesMeshes } from '../mesh/planesmeshes.js';
import { Event, Event1 } from '@zappar/zappar/lib/event';
const raycaster = new THREE.Raycaster();
const vec = new THREE.Vector3();
const matrix = new THREE.Matrix4();
const plane = new THREE.Plane();
const placeButtonDOM = `
<button class="zappar-three-place-ui-button">
	<div id="zappar-three-place-ui-button-icon">
		<svg xmlns="http://www.w3.org/2000/svg" height="30px" viewBox="0 -960 960 960" width="30px" fill="#ffffff"><path d="M379.33-244 154-469.33 201.67-517l177.66 177.67 378.34-378.34L805.33-670l-426 426Z"/></svg>
	</div>
	<div class="zappar-three-place-ui-button-text"></div>
</button>
`;
const resetButtonDOM = `
<button class="zappar-three-place-ui-button">
	<div id="zappar-three-place-ui-button-icon">
	<svg xmlns="http://www.w3.org/2000/svg" height="30px" viewBox="0 -960 960 960" width="30px" fill="#ffffff"><path d="M440-280q-7 0-12-4t-7-10q-14-42-34-70t-40-54q-20-26-33.5-54T300-540q0-58 41-99t99-41q58 0 99 41t41 99q0 40-13.5 68T533-418q-20 26-40 54t-34 70q-2 6-7 10t-12 4Zm0-112q9-14 18-26t17-23q23-30 34-50t11-49q0-33-23.5-56.5T440-620q-33 0-56.5 23.5T360-540q0 29 11 49t34 50q8 11 17 23t18 26Zm0-98q21 0 35.5-14.5T490-540q0-21-14.5-35.5T440-590q-21 0-35.5 14.5T390-540q0 21 14.5 35.5T440-490Zm2 372.67q-151.33 0-257-105.67Q79.33-328.67 79.33-480q0-75.67 28.5-141.5t77.67-115q49.17-49.17 115-77.67t141.5-28.5q75.67 0 141.5 28.5t115 77.67q49.17 49.17 77.67 115t28.5 141.5v14.67L869-530l47 46.67-144.67 144.66-144.66-144.66 47-46.67L738-465.67V-480q0-122-87-209t-209-87q-122 0-209 87t-87 209q.33 122 87.17 209Q320-184 442-184q57 0 109.33-21.5Q603.67-227 643-264l47.67 47.67q-48.34 46-113.17 72.5T442-117.33ZM440-540Z"/></svg>
	</div>
	<div class="zappar-three-place-ui-button-text"></div>
</button>
`;
export const css = `

#overlay {
    position: fixed;
    top: 0;
    left: 0;
    pointer-events: none;
    overflow: hidden;
    z-index: 1000000;
    width: 100%;
    height: 100%;
}

#overlay > * {
    pointer-events: auto;
    position: absolute;
}

.zappar-three-place-ui {
	position: absolute;
	pointer-events: none;
	display: flex;
	flex-direction: column;
	padding: 40px;
	box-sizing: border-box;
}

.zappar-three-place-ui-button {
	outline: none;
	background-color: #555;
	border-radius: 40px;
	font-family: sans-serif;
	color: white;
	font-size: 16px;
	border: none;
	min-width: 60px;
	height: 60px;
	display: flex;
    align-items: center;
    padding-left: 15px;
    padding-right: 15px;
    justify-content: center;
}

.zappar-three-place-ui-button:disabled {
	opacity: 0.1;
}

.zappar-three-place-ui-button-text {
	padding-left: 7px;
	padding-right: 7px;
}

.zappar-three-place-ui-horizontal-left {
	align-items: flex-start;
}

.zappar-three-place-ui-horizontal-right {
	align-items: flex-end;
}

.zappar-three-place-ui-horizontal-center {
	align-items: center;
}

.zappar-three-place-ui-vertical-top {
	top: 0px;
}

.zappar-three-place-ui-vertical-bottom {
	bottom: 0px;
}

.zappar-three-place-ui-vertical-center {
	top: 50%;
	margin-top: -30px;
}`;
let cssloaded = false;
function requireCSS() {
    if (cssloaded)
        return;
    const link = document.createElement('link');
    link.setAttribute('rel', 'stylesheet');
    // @ts-ignore
    link.href = `data:text/css,${encodeURIComponent(css)}`;
    if (document.head)
        document.head.appendChild(link);
    else if (document.body)
        document.body.appendChild(link);
    cssloaded = true;
}
const overlay = document.createElement('div');
overlay.id = 'overlay';
/**
 * A THREE.Group that lets users place AR content in their environment.
 * Provides touch interaction and UI for positioning content on detected surfaces or ground plane.
 * Content can be placed once or repositioned using tap, with optional long-press to move functionality.
 */
export class UserPlacementAnchorGroup extends THREE.Group {
    /**
     * Controls if content is in placement mode
     * @default false
     */
    get placementMode() {
        return this._placementMode;
    }
    /**
     * Controls if content is in placement mode
     * @default false
     */
    set placementMode(value) {
        if (value !== this._placementMode) {
            this._placementMode = value;
            if (!this.worldTracker.enabled)
                return;
            if (value)
                this.onPlacementStart.emit();
            else
                this.onPlacementEnd.emit();
        }
    }
    /**
     * CSS classes for the placement UI container div
     */
    set placeUIClassList(value) {
        this._placeUIClassList = value;
        this.updateUI();
    }
    get placeUIClassList() {
        return this._placeUIClassList;
    }
    /**
     * Gets/sets the enabled state of the world tracker. Disable when not in use to save computational resources during frame processing.
     */
    set enabled(value) {
        this.worldTracker.enabled = value;
    }
    /**
     * Gets/sets the enabled state of the world tracker. Disable when not in use to save computational resources during frame processing.
     */
    get enabled() {
        return this.worldTracker.enabled;
    }
    constructor(camera, worldTracker, renderer, planeMaterial) {
        var _a;
        super();
        this.camera = camera;
        this.worldTracker = worldTracker;
        this.renderer = renderer;
        this._placementMode = false;
        this._placeUIClassList = ['zappar-three-place-ui'];
        /** Indicates if system is ready for user interaction */
        this.ready = false;
        /**
         * Allow placement on detected real-world surfaces
         * @default true
         */
        this.placeOnPlanes = true;
        /**
         * Allow placement on detected ground plane
         * @default false
         */
        this.placeOnGround = false;
        /**
         * Show detected surfaces during placement
         * @default true
         */
        this.visualizePlanesDuringPlacement = true;
        /**
         * Show content during placement
         * @default true
         */
        this.showContentDuringPlacement = true;
        /**
         * Show placement confirmation button
         * Located in #overlay > .zappar-three-place-ui > .zappar-three-place-ui-button
         * @default true
         */
        this.showPlaceButton = true;
        /**
         * Show checkmark icon in place button
         * Located in #overlay > .zappar-three-place-ui > .zappar-three-place-ui-button > #zappar-three-place-ui-button-icon
         * @default true
         */
        this.showPlaceButtonIcon = true;
        /**
         * Custom text for place button
         * Located in #overlay > .zappar-three-place-ui > .zappar-three-place-ui-button > .zappar-three-place-ui-button-text
         * @default ""
         */
        this.placeButtonText = '';
        /**
         * Show button to restart placement
         * Located in #overlay > .zappar-three-place-ui > .zappar-three-place-ui-button
         * @default true
         */
        this.showRestartPlacementButton = true;
        /**
         * Show reset icon in restart button
         * Located in #overlay > .zappar-three-place-ui > .zappar-three-place-ui-button > #zappar-three-place-ui-button-icon
         * @default true
         */
        this.showRestartPlacementButtonIcon = true;
        /**
         * Custom text for restart button
         * Located in #overlay > .zappar-three-place-ui > .zappar-three-place-ui-button > .zappar-three-place-ui-button-text
         * @default ""
         */
        this.restartPlacementButtonText = '';
        /**
         * Long press behavior: none, toggle, or hold
         * @default "none"
         */
        this.longPressToMove = 'none';
        /**
         * Show long press instruction after placement
         * Appears as overlay text in #overlay center
         * @default false
         */
        this.showLongPressInstructionOnPlace = false;
        /**
         * Show long press instruction after placement
         * Appears as overlay text in #overlay center
         * @default false
         */
        this.longPressInstruction = 'Tap and hold to move';
        /**
         * Place button horizontal position
         * Adds .zappar-three-place-ui-horizontal-[left|center|right] class
         * @default "center"
         */
        this.placeButtonHorizontalAlignment = 'center';
        /**
         * Place button vertical position
         * Adds .zappar-three-place-ui-vertical-[top|center|bottom] class
         * @default "bottom"
         */
        this.placeButtonVerticalAlignment = 'bottom';
        /**
         * Restart button horizontal position
         * Adds .zappar-three-place-ui-horizontal-[left|center|right] class
         * @default "left"
         */
        this.restartPlacementButtonHorizontalAlignment = 'left';
        /**
         * Restart button vertical position
         * Adds .zappar-three-place-ui-vertical-[top|center|bottom] class
         * @default "bottom"
         */
        this.restartPlacementButtonVerticalAlignment = 'bottom';
        this.fingerHeld = false;
        this.hasFirstPlaced = false;
        this.placementValid = false;
        this.contentGroup = new THREE.Group();
        /**
         * Emits when placement mode starts
         */
        this.onPlacementStart = new Event();
        /**
         * Emits when placement mode ends
         */
        this.onPlacementEnd = new Event();
        /**
         * Emits when a plane is first intersected.
         */
        this.onPlaneIntersect = new Event1();
        /**
         * Emits when a plane is no longer intersected.
         */
        this.onPlaneIntersectEnd = new Event1();
        this.onQuality = new Zappar.Event1();
        this.handleTouchStart = (evt) => {
            const canvas = this.renderer.domElement;
            if (this.longPressToMove === 'none')
                return;
            let cancelled = false;
            setTimeout(() => {
                if (cancelled)
                    return;
                if (this.longPressToMove === 'hold')
                    this.fingerHeld = true;
                this.placementMode = true;
                this.updateUI();
            }, 1000);
            if (this.longPressToMove === 'hold') {
                const touchEnd = () => {
                    cancelled = true;
                    this.fingerHeld = false;
                    this.placementMode = false;
                    this.updateUI();
                    canvas.removeEventListener('touchend', touchEnd);
                };
                canvas.addEventListener('touchend', touchEnd);
            }
        };
        this.handleTouchEnd = () => {
            if (this.longPressToMove === 'hold') {
                this.fingerHeld = false;
                this.placementMode = false;
                this.updateUI();
            }
        };
        this.onFrameUpdate = () => {
            var _a, _b;
            if (!this.worldTracker.enabled)
                return;
            if (this.camera.pipeline.frameNumber() < 1)
                return;
            this.updateReadyStatus();
            this.planes.visible = this.visualizePlanesDuringPlacement && this.placementMode;
            if (!this.placementMode) {
                if (this.anchor.status === Zappar.AnchorStatus.ANCHOR_STATUS_STOPPED) {
                    for (const plane of this.worldTracker.planes.values()) {
                        // eslint-disable-next-line no-continue
                        if (!plane.firstPlane)
                            continue;
                        this.anchor.setPoseFromAnchorOffset(plane, 0, 0, 0, Zappar.TransformOrientation.Z_TOWARDS_CAMERA);
                        this.contentGroup.visible = true;
                        this.handleFirstPlaced();
                        this.updateUI();
                        return;
                    }
                }
                else {
                    this.contentGroup.matrix.fromArray(this.anchor.pose(this.camera.rawPose, this.camera.currentMirrorMode === CameraMirrorMode.Poses));
                    this.contentGroup.matrix.decompose(this.contentGroup.position, this.contentGroup.quaternion, this.contentGroup.scale);
                    this.contentGroup.updateMatrixWorld();
                    this.placementValid = true;
                    this.contentGroup.visible = true;
                    this.updateUI();
                }
                return;
            }
            // We're in placement mode
            this.camera.updateMatrixWorld();
            raycaster.setFromCamera(new THREE.Vector2(0, -0.3), this.camera);
            let intersection;
            if (this.placeOnPlanes) {
                const planeIntersections = this.planes.intersect(raycaster);
                if (planeIntersections.length > 0) {
                    const local = new THREE.Vector3();
                    local.copy(planeIntersections[0].intersection.point);
                    planeIntersections[0].intersection.object.worldToLocal(local);
                    intersection = [planeIntersections[0].anchorId, planeIntersections[0].intersection.point, local, planeIntersections[0].intersection.distance];
                }
            }
            if (!intersection && this.placeOnGround && ((_b = (_a = this.worldTracker) === null || _a === void 0 ? void 0 : _a.groundAnchor) === null || _b === void 0 ? void 0 : _b.status) !== 0) {
                matrix.fromArray(this.worldTracker.groundAnchor.pose(this.camera.rawPose, this.camera.currentMirrorMode === CameraMirrorMode.Poses));
                vec.set(0, 1, 0);
                plane.set(vec, 0);
                plane.applyMatrix4(matrix);
                const distance = raycaster.ray.distanceToPlane(plane);
                if (distance !== null && (!intersection || intersection[3] > distance)) {
                    const point = new THREE.Vector3();
                    const result = raycaster.ray.intersectPlane(plane, point);
                    if (result) {
                        const local = new THREE.Vector3();
                        matrix.invert();
                        local.copy(point);
                        local.applyMatrix4(matrix);
                        intersection = [this.worldTracker.groundAnchor.id, point, local, distance];
                    }
                }
            }
            if (!intersection) {
                this.placementValid = false;
                this.contentGroup.visible = false;
            }
            else {
                this.placementValid = true;
                this.contentGroup.visible = this.showContentDuringPlacement;
                this.anchor.setPoseFromAnchorOffset(intersection[0], intersection[2].x, intersection[2].y, intersection[2].z, Zappar.TransformOrientation.Z_TOWARDS_CAMERA);
                this.contentGroup.matrix.fromArray(this.anchor.pose(this.camera.rawPose, this.camera.currentMirrorMode === CameraMirrorMode.Poses));
                this.contentGroup.matrix.decompose(this.contentGroup.position, this.contentGroup.quaternion, this.contentGroup.scale);
                this.contentGroup.updateMatrixWorld();
            }
            this.updateUI();
        };
        this.updateReadyStatus = () => {
            this.ready = this.isReady();
        };
        requireCSS();
        this.anchor = new Zappar.CustomAnchor(this.worldTracker);
        this.planes = new PlanesMeshes(camera, worldTracker, planeMaterial);
        this.planes.onIntersect.bind(intersection => {
            if (!this.enabled)
                return;
            this.onPlaneIntersect.emit(intersection);
        });
        this.planes.onIntersectEnd.bind(intersection => {
            if (!this.enabled)
                return;
            this.onPlaneIntersectEnd.emit(intersection);
        });
        this.add(this.planes);
        camera.onCameraPoseUpdate.bind(() => {
            this.onFrameUpdate();
        });
        (_a = renderer.domElement.parentElement) === null || _a === void 0 ? void 0 : _a.appendChild(overlay);
        this.add(this.contentGroup);
        this.contentGroup.visible = false;
        this.add = (...args) => {
            this.contentGroup.add(...args);
            return this;
        };
        this.remove = (...args) => {
            this.contentGroup.remove(...args);
            return this;
        };
        if (this.worldTracker.enabled)
            this.onQuality.emit(worldTracker.quality);
        this.worldTracker.onQuality.bind(quality => {
            if (this.worldTracker.enabled)
                this.onQuality.emit(quality);
        });
        this.setupTouchListeners();
    }
    setupTouchListeners() {
        const canvas = this.renderer.domElement;
        canvas.addEventListener('touchstart', this.handleTouchStart);
        canvas.addEventListener('touchend', this.handleTouchEnd);
    }
    handleFirstPlaced() {
        if (this.hasFirstPlaced)
            return;
        this.hasFirstPlaced = true;
        if (this.showLongPressInstructionOnPlace && this.longPressToMove !== 'none') {
            this.showTextAlert(this.longPressInstruction);
        }
    }
    updateUI() {
        const isPlacing = this.placementMode;
        if (isPlacing && this.showPlaceButton && !this.fingerHeld && this.ready) {
            this.updatePlaceButton();
        }
        else {
            this.removePlaceButton();
        }
        if (!isPlacing && this.showRestartPlacementButton && this.ready) {
            this.updateRestartPlacementButton();
        }
        else {
            this.removeRestartPlacementButton();
        }
    }
    updatePlaceButton() {
        var _a;
        if (!this.placeUI) {
            this.placeUI = document.createElement('div');
            this.placeUI.innerHTML = placeButtonDOM;
            this.placeButton = this.placeUI.querySelector('.zappar-three-place-ui-button');
            overlay.append(this.placeUI);
            (_a = this.placeButton) === null || _a === void 0 ? void 0 : _a.addEventListener('click', () => {
                this.placementMode = false;
            });
        }
        this.placeUI.className = this.placeUIClassList.join(' ');
        const iconElement = this.placeUI.querySelector('#zappar-three-place-ui-button-icon');
        if (iconElement) {
            iconElement.style.display = this.showPlaceButtonIcon ? 'block' : 'none';
        }
        const textElement = this.placeUI.querySelector('.zappar-three-place-ui-button-text');
        if (textElement) {
            textElement.textContent = this.placeButtonText;
            textElement.style.display = this.placeButtonText ? 'block' : 'none';
        }
        this.setElementPosition(this.placeUI, this.placeButtonHorizontalAlignment, this.placeButtonVerticalAlignment);
        if (this.placeButton)
            this.placeButton.disabled = !this.placementValid;
    }
    updateRestartPlacementButton() {
        var _a;
        if (!this.restartPlacementUI) {
            this.restartPlacementUI = document.createElement('div');
            this.restartPlacementUI.innerHTML = resetButtonDOM;
            this.restartPlacementButton = this.restartPlacementUI.querySelector('.zappar-three-place-ui-button');
            overlay.append(this.restartPlacementUI);
            (_a = this.restartPlacementButton) === null || _a === void 0 ? void 0 : _a.addEventListener('click', () => {
                this.placementMode = true;
            });
        }
        this.restartPlacementUI.className = this.placeUIClassList.join(' ');
        const iconElement = this.restartPlacementUI.querySelector('#zappar-three-place-ui-button-icon');
        if (iconElement) {
            iconElement.style.display = this.showRestartPlacementButtonIcon ? 'block' : 'none';
        }
        const textElement = this.restartPlacementUI.querySelector('.zappar-three-place-ui-button-text');
        if (textElement) {
            textElement.textContent = this.restartPlacementButtonText;
            textElement.style.display = this.restartPlacementButtonText ? 'block' : 'none';
        }
        this.setElementPosition(this.restartPlacementUI, this.restartPlacementButtonHorizontalAlignment, this.restartPlacementButtonVerticalAlignment);
    }
    removePlaceButton() {
        if (this.placeUI) {
            this.placeUI.remove();
            this.placeUI = undefined;
            this.placeButton = undefined;
        }
    }
    removeRestartPlacementButton() {
        if (this.restartPlacementUI) {
            this.restartPlacementUI.remove();
            this.restartPlacementUI = undefined;
            this.restartPlacementButton = undefined;
        }
    }
    // eslint-disable-next-line class-methods-use-this
    setElementPosition(element, horizontalAlignment, verticalAlignment) {
        element.style.position = 'absolute';
        switch (horizontalAlignment) {
            case 'left':
                element.style.left = '10px';
                break;
            case 'right':
                element.style.right = '10px';
                break;
            case 'center':
                element.style.left = '50%';
                element.style.transform = 'translateX(-50%)';
                break;
            default:
                console.error('Invalid horizontal alignment');
                break;
        }
        switch (verticalAlignment) {
            case 'top':
                element.style.top = '10px';
                break;
            case 'bottom':
                element.style.bottom = '10px';
                break;
            case 'center':
                element.style.top = '50%';
                element.style.transform += ' translateY(-50%)';
                break;
            default:
                console.error('Invalid vertical alignment');
                break;
        }
    }
    isReady() {
        if (!this.worldTracker)
            return false;
        if (this.worldTracker.quality === Zappar.WorldTrackerQuality.WORLD_TRACKER_QUALITY_INITIALIZING)
            return false;
        if (this.anchor.status === Zappar.AnchorStatus.ANCHOR_STATUS_STOPPED)
            return false;
        return true;
    }
    // eslint-disable-next-line class-methods-use-this
    showTextAlert(text) {
        const alertElement = document.createElement('div');
        alertElement.textContent = text;
        alertElement.style.position = 'absolute';
        alertElement.style.top = '50%';
        alertElement.style.left = '50%';
        alertElement.style.transform = 'translate(-50%, -50%)';
        alertElement.style.backgroundColor = 'rgba(0, 0, 0, 0.7)';
        alertElement.style.color = 'white';
        alertElement.style.padding = '10px';
        alertElement.style.borderRadius = '5px';
        overlay.appendChild(alertElement);
        setTimeout(() => alertElement.remove(), 3000);
    }
    dispose() {
        this.anchor.destroy();
        this.removePlaceButton();
        this.removeRestartPlacementButton();
        this.renderer.domElement.removeEventListener('touchstart', this.handleTouchStart);
        this.renderer.domElement.removeEventListener('touchend', this.handleTouchEnd);
    }
}
