// @zappar/zappar-threejs/lib/trackers/worldplacementgroup.js
/* eslint-disable no-underscore-dangle */
import * as THREE from 'three';
const _ray = new THREE.Ray();
const _up = new THREE.Vector3(0, 1, 0);
const _groundPlane = new THREE.Plane(_up);
const _matrix = new THREE.Matrix4();
const _dest = new THREE.Vector3(0, 0, 0);
const _point = new THREE.Vector3(0, 1, 0);
const _pos = new THREE.Vector3(0, 0, 0);
/**
 * The position of content placed in this group can be chosen by the user.
 *
 *  This class is not a tracker, but a group that can be added to a tracker group.
 */
export class WorldPlacementGroup extends THREE.Group {
    constructor(camera) {
        super();
        this.camera = camera;
        /**
         * While placing, the group will be rotated so the positive `z` axis points towards the camera.
         */
        this.faceCamera = true;
        /**
         * When `true` at runtime, the position of the contents of this group will be reset every frame to a point directly in front of the camera.
         */
        this.placementMode = false;
        /**
         * By default, with this parameter `false` the scale of the contents of this group is that of the parent group.
         * If set to `true`, this group will adjust the scale based on the height of the camera above the origin.
         */
        this.normalizeScale = false;
    }
    /**
     * Called internally by ground anchor group.
     */
    update() {
        if (this.placementMode)
            this.groundPlaneIntersection();
    }
    groundPlaneIntersection() {
        if (!this.parent)
            return;
        _up.set(0, 1, 0);
        _groundPlane.set(_up, 0);
        _groundPlane.applyMatrix4(this.parent.matrixWorld);
        this.camera.updateMatrixWorld();
        _ray.origin.set(0, 0, 0);
        _ray.direction.set(0, 0, -1);
        _ray.applyMatrix4(this.camera.matrixWorld);
        if (!_ray.intersectPlane(_groundPlane, _dest))
            return;
        _matrix.copy(this.parent.matrixWorld);
        _matrix.invert();
        _dest.applyMatrix4(_matrix);
        if (Number.isNaN(_dest.x) || Number.isNaN(_dest.y) || Number.isNaN(_dest.z))
            return;
        if (this.faceCamera) {
            this.camera.getWorldPosition(_pos);
            this.rotation.y = Math.atan2(_pos.x - _dest.x, _pos.z - _dest.z);
            this.position.copy(_dest);
        }
        if (this.normalizeScale) {
            _matrix.multiply(this.camera.matrixWorld);
            _point.set(_matrix.elements[12], _matrix.elements[13], _matrix.elements[14]);
            const cameraHeight = _groundPlane.distanceToPoint(_point);
            if (cameraHeight > 0)
                this.scale.set(cameraHeight, cameraHeight, cameraHeight);
        }
        this.updateMatrix();
        this.matrixWorldNeedsUpdate = true;
        this.updateMatrixWorld(true);
    }
}
