// @zappar/zappar-threejs/lib/trackers/worldtrackinganchorgroup.js
import { AnchorStatus } from '@zappar/zappar';
import * as THREE from 'three';
import { CameraMirrorMode } from '../camera.js';
/**
 * A THREE.Group which attaches content to a point in the world as it moves around in the camera view.
 * @see https://docs.zap.works/universal-ar/web-libraries/threejs/world-tracking/
 */
export class WorldTrackingAnchorGroup extends THREE.Group {
    /**
     * Constructs a new WorldTrackingAnchorGroup.
     * @param camera - A ZapparThree.Camera.
     * @param worldTracker - The world tracker which will be used.
     */
    constructor(camera, worldTracker) {
        super();
        this.camera = camera;
        this.worldTracker = worldTracker;
        this.status = AnchorStatus.ANCHOR_STATUS_INITIALIZING;
        // We'll be updating the inner group's matrix ourselves from the Zappar library
        this.matrixAutoUpdate = false;
    }
    updatePose() {
        if (this.anchor) {
            this.status = this.anchor.status;
            this.matrix.fromArray(this.anchor.pose(this.camera.rawPose, this.camera.currentMirrorMode === CameraMirrorMode.Poses));
            this.matrix.decompose(this.position, this.quaternion, this.scale);
            // this.innerGroup.updateMatrixWorld();
        }
    }
    updateMatrixWorld(force) {
        this.updatePose();
        if (!this.anchor && this.status !== AnchorStatus.ANCHOR_STATUS_INITIALIZING) {
            this.status = AnchorStatus.ANCHOR_STATUS_STOPPED;
        }
        switch (this.status) {
            // ANCHOR_STATUS_INITIALIZING?
            case AnchorStatus.ANCHOR_STATUS_PAUSED:
            case AnchorStatus.ANCHOR_STATUS_STOPPED:
                // this.innerGroup.visible = false;
                break;
            case AnchorStatus.ANCHOR_STATUS_TRACKING:
                // this.innerGroup.visible = true;
                break;
            default:
                console.warn(`Unknown anchor status: ${this.status}`);
        }
        super.updateMatrixWorld(force);
    }
    dispose() {
        this.worldTracker.destroy();
    }
}
